drivers: i2c: document "slave" API

pabigot opened the pull request:
Document the API used for I2C slave devices.  Fill out the peripheral documentation with a reference to the I2C specification, a short summary of the roles of I2C devices, and a note on language used.  Remove some stubs from drivers that don't implement the slave API.  The focus of the PR is documenting the existing following API, not documenting the I2C subsystem as a whole.

The PR includes a change that distinguishes the slave API as Experimental in the API stability table, as it is not satisfactory and has limited support.  See also #27675.

The documented behavior is derived primarily from the LPC11U6X implementation, as the STM32 implementation generally ignores return values.  There may be errors in the description, and there are certainly gaps in the existing driver implementations.

Updated 2021-02-06 to revert to "master/slave" terminology in conformance to [the coding guidelines](https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html?highlight=inclusive#rule-a-2-inclusive-language).

Fixes #27879
References #27033

pabigot committed:
drivers: i2c: litex: remove redundant null pointers

The callback pointers for uninitiated operations are implicitly null;
making them explicit only confuses maintainers searching for drivers
that implement the API.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
doc: drivers: i2c: provide more information on I2C APIs

Although the master API is stable, the slave API has never been
documented and has only two in-tree implementations with no uses
outside of a single test application.  It is marked experimental in
this commit.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: i2c: document slave API

The I2C slave API has been present since 1.12 but lacked documentation
on the behavior of its functions.  Provide that, and revise existing
documentation to make clear the type and mode of the device passed to
each API call.

Documented behavior is derived primarily from the LPC11U6X
implementation, as the STM32 implementation generally ignores return
values.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

olofj commented:
I don't think the "leader" and "follower" terminology makes sense in the i2c context -- the i2c target/device isn't "following" the leader in the same way that a main/replica database would do, for example.

"i2c controller" or "i2c host" or "i2c initiator" and "i2c device" or "i2c target" seem like more appropriate terminology here, unless leader/follower has already gained significant use elsewhere. I think this is also consistent with some of the feedback on #27033.

franciscomunoz reviewed (dismissed):
Overall looks good. I don't have anything to comment about the terms master/slave.

pabigot commented:
@olofj I've added https://github.com/zephyrproject-rtos/zephyr/issues/27033#issuecomment-686405572 to further explain why I'm using these terms.  Please provide further feedback on the name selection in that PR.

nashif commented on doc/reference/peripherals/i2c.rst:
```diff
@@ -1,12 +1,58 @@
 
+.. note::
+
+   Zephyr recognizes the need to change the racially-charged terms
```
IMO this type of note should be done in one place in the Zephyr documentation and not per peripheral.

nashif commented:
> @olofj I've added [#27033 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/27033#issuecomment-686405572) to further explain why I'm using these terms. Please provide further feedback on the name selection in that PR.

We should probably consider splitting this PR and not mix a documentation bug fix with inclusive language changes that still being discussed.

pabigot commented:
The documentation requires use of three distinguishing terms, currently "controller", "leader", and "follower".  We can remove the explanation of why those terms are used if necessary.

I would not be happy about being forced to change the documentation to continue to use "master" and "slave".

nashif commented:
> I would not be happy about being forced to change the documentation to continue to use "master" and "slave".

making this change here and leaving i2c with slave/master usage especially in with the APIs still using slave/master in their function names is not going to make anyone happy. Once we know and agree on what terms we should be using this should be done across the tree, not only for the "slave".

pabigot commented:
At this time no changes have been requested, just comments that express opinions about the use of "leader" and "follower".  My opinion differs, or I'd have done something else.  My detailed responses are at/around https://github.com/zephyrproject-rtos/zephyr/issues/27033#issuecomment-686405572.

Here I would still like reviews related to the technical content, so we can move forward on documenting this API.

nashif commented:
Can we move this forward based on the discussion we had 2 weeks ago re inclusive language please?

pabigot commented:
Waiting on this action item for [the public decision on how Zephyr will be addressing non-inclusive language](https://github.com/zephyrproject-rtos/zephyr/issues/27033#issuecomment-687974083).

>Transitioning Subsystems
>    Create page on documentation for transition, including table with status of subsystems
>        Skeleton for the page - Maureen
>        Subsystem maintainers own updating status / plans

nashif commented:
> Waiting on this action item for [the decision on how Zephyr will be addressing non-inclusive language](https://github.com/zephyrproject-rtos/zephyr/issues/27033#issuecomment-687974083).

@MaureenHelm  any updates?

MaureenHelm commented:
> > Waiting on this action item for [the decision on how Zephyr will be addressing non-inclusive language](https://github.com/zephyrproject-rtos/zephyr/issues/27033#issuecomment-687974083).
> 
> @MaureenHelm any updates?

Working on a PR now.

pabigot committed:
doc: coding guidelines: add an anchor for inclusive language guideline

Documentation for subsystems that use offensive terms by standard may
need to reference this section as justification for their continued
and new use.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: i2c_slave_api: update to current standard terminology

Zephyr does not currently allow deviation from standard terminology
for a technology even if it is non-inclusive, until the corresponding
standards body has confirmed intent to change that terminology.  The
terms used in a previous attempt to be inclusive do not match the
expected forthcoming standard terms.

Revert to standard terms until the new ones have been announced and
the switch made throughout Zephyr.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
doc: inclusive language: reference I2C status issue

Status of addressing inclusive language concerns is embedded into the
coding guidelines.  Add a link to the corresponding issue for I2C.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nashif commented:
isnt this a documentation change? Why should this not go into 2.5?

pabigot commented:
If you want it there, sure.  I was cleaning up the PRs I'd wanted to get in to this version, and this wasn't specifically one of them because of the delay in getting the policy in place.  Feel free to add it back if you intend to merge it.

pabigot commented:
Ah, my mistake, I thought this was a milestone I'd assigned.  Back now.
